File: templating/src/main/python/jsonteng/rules/one_of_rule.py
# Copyright 2019 VMware, Inc.
# SPDX-License-Indentifier: Apache-2.0

from ..exception import TemplateEngineException
from .rule_base import RuleBase


class OneOfRule(RuleBase):
    """
    Select a value from a list of condition/value pairs if the first condition
    is true. If the last element contains value only, return it if all other
    conditions are false.
    """
    name = "one-of"

    def __init__(self, rule_resolver):
        """
        Construct this rule.
        :param rule_resolver: Rule resolver
        :type rule_resolver: 'RuleResolver'
        """
        super().__init__(rule_resolver)
        self._element_resolver = rule_resolver.get_element_resolver()

    def process(self, rule_tokens, binding_data_list):
        """
        Process this rule.
        :param rule_tokens: Rule arguments.
        :type rule_tokens: 'list'
        :param binding_data_list: Binding data used during the processing.
        :type binding_data_list: 'list'
        :return: JSON object
        :rtype: JSON object
        """
        token_count = len(rule_tokens)
        if token_count < 1:
            raise TemplateEngineException(
                "Rule \"{}\" needs at least 1 parameter."
                " Parameters given {}".format(OneOfRule.name, rule_tokens))
        for index, item in enumerate(rule_tokens):
            if isinstance(item, list) and len(item) == 2:
                condition = item[0]
                value = item[1]
                condition_expr = self._element_resolver.resolve(
                    condition, binding_data_list)
                if self.safe_eval(condition_expr):
                    return self._element_resolver.resolve(
                        value, binding_data_list)
            else:
                if index == (token_count - 1) and not isinstance(item, list):
                    return self._element_resolver.resolve(
                        rule_tokens[-1], binding_data_list)
                else:
                    raise TemplateEngineException(
                        "Rule \"one-of\" contains an invalid parameter."
                        " {}".format(item))
        return RuleBase.RULE_NONE

File: templating/src/main/java/com/vmware/jsonteng/rules/OneOfRule.java
// Copyright 2019 VMware, Inc.
// SPDX-License-Indentifier: Apache-2.0

package com.vmware.jsonteng.rules;

import java.util.List;
import java.util.Map;

import com.vmware.jsonteng.ElementResolver;
import com.vmware.jsonteng.RuleResolver;
import com.vmware.jsonteng.TemplateEngineException;

public class OneOfRule extends RuleBase {
    static final String name = "one-of";

    private final ElementResolver elementResolver;

    public OneOfRule(RuleResolver ruleResolver) {
        super(ruleResolver);
        elementResolver = ruleResolver.getElementResolver();
    }

    @Override
    public Object process(List<?> ruleTokens, List<Map<String, ?>> bindingDataList) throws TemplateEngineException {
        int tokenCount = ruleTokens.size();
        if (tokenCount < 1) {
            throw new TemplateEngineException(
                    String.format("Rule \"%s\" needs at least 1 parameters. Parameters given %s", name, ruleTokens));
        }
        for (int i = 0; i < tokenCount; i++) {
            Object item = ruleTokens.get(i);
            if (item instanceof List && ((List) item).size() == 2) {
                Object condition = ((List) item).get(0);
                Object value = ((List) item).get(1);
                Object conditionExpr = elementResolver.resolve(condition, bindingDataList);
                boolean result = safeEval(conditionExpr);
                if (result) {
                    return elementResolver.resolve(value, bindingDataList);
                }
            }
            else {
                if (i == (tokenCount -1) && !(item instanceof List)) {
                    return elementResolver.resolve(ruleTokens.get(tokenCount - 1), bindingDataList);
                }
                else {
                    throw new TemplateEngineException(
                            String.format("Rule \"%s\" contains an invalid parameter. %s", name, item));
                }
            }
        }
        return RuleBase.RULE_NONE;
    }
}
